builder/beacon_client.go

```diff
@@ -51,14 +62,54 @@ func (b *BeaconClient) isValidator(pubkey PubkeyHex) bool {
 }
 
 func (b *BeaconClient) getProposerForNextSlot(requestedSlot uint64) (PubkeyHex, error) {
-	/* Only returns proposer if requestedSlot is currentSlot + 1, would be a race otherwise */
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	if b.currentSlot+1 != requestedSlot {
-		return PubkeyHex(""), errors.New("slot out of sync")
+	if (requestedSlot / b.slotsInEpoch) > b.currentEpoch {
+		b.currEpochProposerMap = b.nextEpochProposerMap
+	}
+
+	nextSlotProposer, found := b.currEpochProposerMap[requestedSlot]
+	if !found {
+		log.Error("inconsistent proposer mapping", "requestSlot", requestedSlot, "slotProposerMap", b.currEpochProposerMap)
+		return PubkeyHex(""), errors.New("inconsistent proposer mapping")
+	}
+	return nextSlotProposer, nil
+}
+
+func (b *BeaconClient) updateValidatorsMap() error {
+	// Start regular slot updates
+	slotsPerEpoch := b.slotsInEpoch
+	durationPerSlot := time.Duration(b.secondsInSlot) * time.Second
+	durationPerEpoch := durationPerSlot * time.Duration(slotsPerEpoch)
+	// Every half epoch request validators map
+	timer := time.NewTicker(0)
+	for {
+		select {
+		case <-timer.C:
+			currentSlot, err := fetchCurrentSlot(b.endpoint)
+			if err != nil {
+				log.Error("could not get current slot", "err", err)
+				continue
+			}
+
+			currentEpoch := currentSlot / b.slotsInEpoch
+			if currentEpoch > b.currentEpoch {
```

Comment:
There's actually a race for `b.currentEpoch` between this loop and `onForkchoiceUpdate`
I wonder if just keeping a most up to date slot proposer map instead of two copies would solve most if not all of those issues.
The slot proposer map contains slots from this and the next epochs, so you have a whole epoch of runway to update it (if that fails, there are bigger issues than the builder).
Instead of keeping `nextEpochProposerMap` and `currEpochProposerMap` you could just keep  `currEpochProposerMap` up to date in this loop.